### PR TITLE
C51-49: Integrate case card functionality

### PR DESCRIPTION
**Pull request by swastikpareek**
## Overview
 This PR covers Functionality integration for the case card component on Case list page

## Before
c51-49 - before

## After

c51-49 - after

c51-49 -after selected

## Technical Details
 * For the reusability purpose the HTML is now a part of `civicase-case-card` directive which encapsulates all the functionality there.
 * New files added as a part of this directive
   * `CaseCard.html` - Holds HTML for the card
   * `CaseCard.js` - Holds direcitve and controller info
   * `_case-card.scss` - Holds styles for the card
 * HTML classes changed to BEM structure
   * **civicase__case-list-view** - changed from **civicase-view-with-search** (Used for caselist right view section when case selected)
 * `getActivityFeedUrl` Service dependency moved from `CaseListTable` controller to `Case Card` controller
 * Changes in `contactCardDirective`
   * Added a wrapper HTML div (to have `civicase__contact-card` class as a part of directive itself).
   * Added `replace: true` parameter to the directive to have the wrapper included as part of the directive.
   * Add some styles to fix regression activity card
 * Moved Placeholder loading styles from general.css and moved it to `_loading-placeholder.scss` and changed class to follow BEM.
   * **civicase__loading-placeholder__icon** : changed from `civicase-icon-placeholder` for icon placeholder
   * **civicase__loading-placeholder__activity-card** : changed from `activity-card-loading-placeholder` for activity card loading state
   * **civicase__loading-placeholder--big**: changed from `placeholder-big`
   * **civicase__loading-placeholder__online**: changed from `oneline-loading-placeholder`
   * **civicase__loading-placeholder__case-view-panel**: changed from `case-view-placeholder-panel` for case-view panel loading state
   * **civicase__loading-placeholder__online-strip**: chnaged from `oneline-loading-placeholder-strip`
 * New color variable `$placeholder-grey` added for loading placeholders.
 * Add Moment.js script in `packages/moment.min.js` added loads it in `civicase.php` `hook_civicrm_pageRun()` function.
 * Add `moment` as a global variable in semistandard conf in `package.json`
 * Add `$case-card-width` variable to have single source of truth for case card width.
 * Fix right section of the case card (when one is selcted) - Fix its width and spacing
 * Directives name for `stickyTableHeader` is changed to `civicaseStickyTableHeader` and `sticktFooterPager` is changed to `civicaseStickyFooterPager`.

### Updates
 * Improved logic (used loadash functions)
 * Add documentations
 * Write tests for the case-card directive.
 * Update width of the case card to `400px` to contain counts (in case counts are of two digits) and checkbox for bulk action
 * Fix logic for overdue tasks count
 * Add API chain call (Activity API) to the case card.
 * Added accessibility attributes to the checkboxes
 * Added tooltip on next milestone which will show up next milestone's subject or its type
* Fix activity feed logic for overdue activities for task and communications.

### CODE REVIEW

**Comment by deb1990** [comment]
> Clicking on red tasks should open up tasks that are overdue only
are you sure this is working fine?

**Comment by swastikpareek** [comment]
@deb1990 : Functionality for next milestone has been fixed.
However for the link to other counts functionality can't be achieved because of the following reason -

>Every count on the case card needs to be a link to the activity feed. This can be achieved for tasks and comms but not for other count (as it is not a type but is a logical accumulation of activites which are not `tasks` and `communications`) and filter for activity type is a single select so if we want to se the filter logic for other (SET activity type = "File", "Attachments", "Milestones", etc), we can't.
  So to have it we will need to make the necessary changes to the activity filter (making it multi select) and API changes to allow multiselect in the headers and give correct response.

The link to the overdue task (Filter view) can't be implemented with the current functionality too as there is no activity status category for overdue task only.

So as per our discussion we are having the link to counts to point to nothing for now (till the functionality has been added) 
We need to create a new ticket for the functionality to work.

File: ang/civicase/CaseCard.html [hunk 1]
```diff
@@ -1,109 +1,77 @@
...
-              <span class="civicase__contact-name">Andrew Burke</span>
-            </a>
-            <a href="#">
-              <i class="civicase__contact-icon material-icons">person</i>
-              <span class="civicase__contact-name">John Smith</span>
-            </a>
-          </li>
-        </ul>
-      </div>
+    <div ng-if="!$parent.bulkAllowed" class="civicase__case-card-lock">
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply ang/civicase/CaseCard.html]: To show the use that on this element the bulk actions are not allowed. But as per our discussion I think this should be discussed and implemented while implementing bulk actions functionality.
I will add a todo comment for it

File: scss/components/_checkbox.scss [hunk 2]
```diff
@@ -12,14 +12,31 @@
   transform: scale(0.3);
   transition: 0.2s all cubic-bezier(0, 0, 0, 0.4);
   transition-delay: unset;
-  width: 18px;  
+  width: 18px;
+
+  .civicase__checkbox__input {
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply scss/components/_checkbox.scss]: This class is used on <input> element to be toggled (Added input element for accessibility)

File: ang/civicase/CaseListTable.js [hunk 3]
```diff
@@ -244,7 +241,8 @@
           sort: sort.field + ' ' + sort.dir,
           limit: page.size,
           offset: page.size * (page.num - 1)
-        }
+        },
+        'api.Activity.get': {}
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply ang/civicase/CaseListTable.js]: I will add comments. But the API is smart enough to get the correct activity count for each case. so we don't need to pass the case id.

File: ang/civicase/ContactCard.html [hunk 4]
```diff
@@ -1,55 +1,57 @@
-<span ng-if="!isAvatar">
-  <!-- Contact Icons -->
-  <i class="civicase__contact-icon material-icons" ng-if="getContactIconOf(contacts[0].contact_id) === 'Individual'">person</i>
-  <i class="civicase__contact-icon material-icons" ng-if="getContactIconOf(contacts[0].contact_id) === 'Household'">home</i>
-  <i class="civicase__contact-icon material-icons" ng-if="getContactIconOf(contacts[0].contact_id) === 'Organisation'">domain</i>
-  <!-- End - Contact Icons -->
+<div class="civicase__contact-card">
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply ang/civicase/ContactCard.html]: Nope just this.

File: ang/civicase/View.js [hunk 5]
```diff
@@ -292,14 +288,14 @@
...
-  angular.module('civicase').directive('civicaseView', function() {
+  angular.module('civicase').directive('civicaseView', function () {
     return {
       restrict: 'A',
       template:
         '<div class="panel panel-default civicase-view-panel">' +
           '<div class="panel-header" ng-if="item" ng-include="\'~/civicase/CaseViewHeader.html\'"></div>' +
           '<div class="panel-body case-view-body" ng-if="item" ng-include="\'~/civicase/CaseTabs.html\'"></div>' +
-          '<div ng-if="!item" class="case-view-placeholder-panel" ng-include="\'~/civicase/CaseViewPlaceholder.html\'"></div>' +
+          '<div ng-if="!item" class="civicase__loading-placeholder__case-view-panel" ng-include="\'~/civicase/CaseViewPlaceholder.html\'"></div>' +
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply ang/civicase/View.js]: Nope. Just this.

File: scss/components/_case-card.scss [hunk 6]
```diff
@@ -93,3 +101,23 @@
 .civicase__case-card__activity-count-container {
   margin-right: 8px;
 }
+
+.civicase__case-card-lock {
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply scss/components/_case-card.scss]: We need a discussion on it. Till the time I am commenting on this.

File: ang/test/mocks/data/activity-status-types.data.js [hunk 7]
```diff
@@ -0,0 +1,13 @@
+(function () {
+  var module = angular.module('civicase.data');
+
+  CRM.civicase.activityStatusTypes = {
+    'cancelled': [3, 5, 6, 8],
+    'completed': [2],
+    'incomplete': [1, 4, 7, 9, 10]
+  };
+
+  module.constant('ActivityStatusTypes', {
```
Comment (deb1990) [inline ang/test/mocks/data/activity-status-types.data.js]: Should be called `ActivityStatusTypesData`

File: scss/components/_checkbox.scss [hunk 8]
```diff
@@ -12,14 +12,31 @@
...
+  }
 }
 
 .civicase__checkbox--checked {
   color: $crm-grayblue-darker;
   font-size: 24px;
   margin-left: -4px;
   margin-top: -4px;
+  opacity: 0;
+  transition: 0.2s all cubic-bezier(0, 0, 0, 0.4);
```
[the remark this answers is not in the document]
Comment (swastikpareek) [reply scss/components/_checkbox.scss]: Removed it as we don't need this.

**Pull request merged**